// typescript-to-lua/dist/transformation/visitors/return.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createReturnStatement = exports.transformReturnStatement = exports.transformExpressionBodyToReturnStatement = void 0;
const ts = require("typescript");
const lua = require("../../LuaAST");
const assignment_validation_1 = require("../utils/assignment-validation");
const lua_ast_1 = require("../utils/lua-ast");
const scope_1 = require("../utils/scope");
const call_1 = require("./call");
const multi_1 = require("./language-extensions/multi");
const diagnostics_1 = require("../utils/diagnostics");
const typescript_1 = require("../utils/typescript");
function transformExpressionsInReturn(context, node, insideTryCatch) {
    const expressionType = context.checker.getTypeAtLocation(node);
    if (ts.isCallExpression(node)) {
        // $multi(...)
        if ((0, multi_1.isMultiFunctionCall)(context, node)) {
            // Don't allow $multi to be implicitly cast to something other than LuaMultiReturn
            const type = context.checker.getContextualType(node);
            if (type && !(0, multi_1.canBeMultiReturnType)(type)) {
                context.diagnostics.push((0, diagnostics_1.invalidMultiFunctionReturnType)(node));
            }
            let returnValues = (0, call_1.transformArguments)(context, node.arguments);
            if (insideTryCatch) {
                returnValues = [(0, lua_ast_1.wrapInTable)(...returnValues)]; // Wrap results when returning inside try/catch
            }
            return returnValues;
        }
        // Force-wrap LuaMultiReturn when returning inside try/catch
        if (insideTryCatch && (0, multi_1.returnsMultiType)(context, node) && !(0, multi_1.shouldMultiReturnCallBeWrapped)(context, node)) {
            return [(0, lua_ast_1.wrapInTable)(context.transformExpression(node))];
        }
    }
    else if ((0, multi_1.isInMultiReturnFunction)(context, node) && (0, multi_1.isMultiReturnType)(expressionType)) {
        // Unpack objects typed as LuaMultiReturn
        return [(0, lua_ast_1.createUnpackCall)(context, context.transformExpression(node), node)];
    }
    return [context.transformExpression(node)];
}
function transformExpressionBodyToReturnStatement(context, node) {
    const expressions = transformExpressionsInReturn(context, node, false);
    return createReturnStatement(context, expressions, node);
}
exports.transformExpressionBodyToReturnStatement = transformExpressionBodyToReturnStatement;
const transformReturnStatement = (statement, context) => {
    let results;
    if (statement.expression) {
        const expressionType = context.checker.getTypeAtLocation(statement.expression);
        const returnType = context.checker.getContextualType(statement.expression);
        if (returnType) {
            (0, assignment_validation_1.validateAssignment)(context, statement, expressionType, returnType);
        }
        results = transformExpressionsInReturn(context, statement.expression, isInTryCatch(context));
    }
    else {
        // Empty return
        results = [];
    }
    return createReturnStatement(context, results, statement);
};
exports.transformReturnStatement = transformReturnStatement;
function createReturnStatement(context, values, node) {
    const results = [...values];
    if (isInTryCatch(context)) {
        // Bubble up explicit return flag and check if we're inside a try/catch block
        results.unshift(lua.createBooleanLiteral(true));
    }
    else if ((0, typescript_1.isInAsyncFunction)(node)) {
        // Add nil error handler in async function and not in try
        results.unshift(lua.createNilLiteral());
    }
    return lua.createReturnStatement(results, node);
}
exports.createReturnStatement = createReturnStatement;
function isInTryCatch(context) {
    // Check if context is in a try or catch
    let insideTryCatch = false;
    for (const scope of (0, scope_1.walkScopesUp)(context)) {
        scope.functionReturned = true;
        if (scope.type === scope_1.ScopeType.Function) {
            break;
        }
        insideTryCatch = insideTryCatch || scope.type === scope_1.ScopeType.Try || scope.type === scope_1.ScopeType.Catch;
    }
    return insideTryCatch;
}
